allennlp/models/event2mind.py

```diff
@@ -0,0 +1,348 @@
+        start_decoder_input = self._target_embedder(start_predictions)
+        start_decoder_hidden = self._decoder_cell(start_decoder_input, final_encoder_output)
+        start_output_projections = self._output_projection_layer(start_decoder_hidden)
+        start_class_log_probabilities = F.log_softmax(start_output_projections, dim=-1)
+        start_top_log_probabilities, start_predicted_classes = start_class_log_probabilities.topk(k)
+
+        # Set starting values
+        # [(batch_size, k)]
+        log_probabilities.append(start_top_log_probabilities)
+        # [(batch_size, k)]
+        predictions.append(start_predicted_classes)
+        # Set the same hidden state for each element in beam.
+        # (batch_size * k, _decoder_output_dim)
+        decoder_hidden = start_decoder_hidden.\
+            unsqueeze(1).expand(batch_size, k, self._decoder_output_dim).\
+            reshape(batch_size * k, self._decoder_output_dim)
+
+        for timestep in range(num_decoding_steps - 1):
+            decoder_input = self._target_embedder(predictions[-1].reshape(batch_size * k))
+            # reshape(batch_size * k, self._decoder_output_dim)
+            decoder_hidden = self._decoder_cell(decoder_input, decoder_hidden)
+            # (batch_size * k, num_classes)
+            output_projections = self._output_projection_layer(decoder_hidden)
+
+            class_log_probabilities = F.log_softmax(output_projections, dim=-1)
+            # (batch_size * k, k), (batch_size * k, k)
+            top_log_probabilities, predicted_classes = class_log_probabilities.topk(k)
+            # TODO(brendanr): Account for predicted_class == end_symbol explicitly?
+            # TODO(brendanr): Normalize for length?
+            # (batch_size * k, k)
+            expanded_last_log_probabilities = log_probabilities[-1].\
+                unsqueeze(2).\
+                expand(batch_size, k, k).\
+                reshape(batch_size * k, k)
+            summed_top_log_probabilities = top_log_probabilities + expanded_last_log_probabilities
+
+            reshaped_top_log_probabilities = summed_top_log_probabilities.reshape(batch_size, k * k)
+            reshaped_predicted_classes = predicted_classes.reshape(batch_size, k * k)
+            restricted_beam_log_probs, restricted_beam_indices = reshaped_top_log_probabilities.topk(k)
+            # TODO(brendanr): Something about this is weird. restricted_predicted_classes == restricted_beam_indices???
+            restricted_predicted_classes = reshaped_predicted_classes.gather(1, restricted_beam_indices)
+
+            log_probabilities.append(restricted_beam_log_probs)
+            predictions.append(restricted_predicted_classes)
+            backpointer = restricted_beam_indices / k
+            backpointers.append(backpointer)
+            expanded_backpointer = backpointer.unsqueeze(2).expand(batch_size, k, self._decoder_output_dim)
```

Comment:
@DeNeutoy, this is the are I was telling you about. Essentially I'm using these backpointers to ensure that the hidden layer (for the next iteration) is pruned to only include entries corresponding to items still in the beam. It seems like expanding the backpointers to the `_decoder_output_dim` is enough to make the `gather` work. Feel free to poke around, but I'm going to clean this up before requesting a real review. :) Thanks!

Reply:
If this becomes a sticking point, i'd just loop over the batch to get it out of the way. It's always hard to tell what the bottleneck will be in these models - I think in this case, it might not be the beam search, because by nature of batching together the topk states, it's already being batched reasonably efficiently.